src/System.IO.Pipelines/Pipe.cs

```diff
@@ -412,35 +412,60 @@ void IPipeReader.Advance(ReadCursor consumed, ReadCursor examined)
             BufferSegment returnStart = null;
             BufferSegment returnEnd = null;
 
-            int consumedBytes = 0;
-            if (!consumed.IsDefault)
-            {
-                returnStart = _readHead;
-                consumedBytes = ReadCursor.GetLength(returnStart, returnStart.Start, consumed.Segment, consumed.Index);
-
-                returnEnd = consumed.Segment;
-                _readHead = consumed.Segment;
-                _readHead.Start = consumed.Index;
-            }
-
             // Reading commit head shared with writer
             Action continuation = null;
             lock (_sync)
             {
-                var oldLength = _length;
-                _length -= consumedBytes;
+                var examinedEverything = examined.Segment == _commitHead && examined.Index == _commitHeadIndex;
 
-                if (oldLength >= _maximumSizeLow &&
-                    _length < _maximumSizeLow)
+                if (!consumed.IsDefault)
                 {
-                    continuation = _writerAwaitable.Complete();
+                    if (_readHead == null)
+                    {
+                        PipelinesThrowHelper.ThrowInvalidOperationException(ExceptionResource.AdvanceToInvalidCursor);
+                        return;
+                    }
+
+                    returnStart = _readHead;
+                    returnEnd = consumed.Segment;
+
+                    // Check if we crossed _maximumSizeLow and complete backpressure
+                    var consumedBytes = ReadCursor.GetLength(returnStart, returnStart.Start, consumed.Segment, consumed.Index);
+                    var oldLength = _length;
+                    _length -= consumedBytes;
+
+                    if (oldLength >= _maximumSizeLow &&
+                        _length < _maximumSizeLow)
+                    {
+                        continuation = _writerAwaitable.Complete();
+                    }
+
+                    // Check if we consumed entire last segment
+                    // if we are going to return commit head
+                    // we need to check that there is no writing operation that
+                    // might be using tailspace
+                    if (consumed.Index == returnEnd.End &&
```

Comment:
So this only returns consumed.Segment if the consumer reads the entire block? I see how this fixes #1540 (though even then, it would cause ReadAsync/Advance to be called one more time unnecessarily).

The way I'm reading this, I don't see how this would "Return all blocks when reader catches up to writer" when the writer partially fills the last block which I think would be most of the time. If that's a case, that's a wasted 8KB of request/response buffer space used by Kestrel per idle connection over 1.x.

Am I misunderstanding how this code works?